### ISSUE TITLE
resize_to_layout!() does not work with DataAspect

### ISSUE THREAD

**Issue by rongcuid** [MakieOrg/Makie.jl#3654]
- [x] are you running newest version (version from docs) ?
- [x] can you reproduce the bug with a fresh environment ? (`]activate --temp; add Makie`)
- [x] What platform + GPU are you on?, Mac M2

It is implied by documentation (https://docs.makie.org/stable/reference/blocks/axis/#aspect) that `resize_to_layout!` can be made to work with `AxisAspect` and `DataAspect`, but I am unable to do so.

> To make a `GridLayout` aware of aspect ratio constraints, refer to the `Aspect` column or row size setting.

There is no such setting that I can find.

MVP:

```julia
using CairoMakie
set_theme!(backgroundcolor=:gray90)
fig = Figure()
ax = Axis(fig[1, 1], aspect=DataAspect())
lines!(ax, range(1, 4, 4), range(1, 2, 4))
resize_to_layout!(fig)
fig
```

Generates this plot with large margin:

![plot_46](https://github.com/MakieOrg/Makie.jl/assets/1312525/7da45020-960b-438e-b240-743818b0f77d)

https://github.com/MakieOrg/Makie.jl/issues/1523 is related, but it requires a fixed size, which is not really known as the data is procedurally generated.

**Comment by jkrumbiegel** [comment]
That's not a bug but expected behavior, have a look at this tutorial for an explanation https://docs.makie.org/stable/tutorials/aspect-tutorial/

**Comment by avonmoll** [comment]
I feel that this is a major shortcoming of Makie at this point. I make a lot of plots with `aspect=DataAspect()` and it is a _tremendous_ pain to get the plots ready to include in a publication. The main difficulty is in "cropping" out the region of the figure that is unused by the axis. The MWE provided by the OP would be ideal if it worked (and it seems completely sensible based on the description/docs of `resize_to_layout!`). I tried and tried to automate the generation of a figure like OP's to use `colsize!` and `rowsize!` as described in the linked tutorial, but it feels like a total hack. This kind of functionality ought to be available within Makie itself.

For what it's worth, below is what I came up with. The goal is to get a figure that is at most 3.8" wide and at most 3" tall with `aspect=DataAspect()`. Thus, if you have a plot that is very short and wide it should end up as a figure that is 3.8" wide and <3" tall. If I were to simply call `save`, then, as OP points out, I'd get a figure 3.8" wide and 3" tall which would waste space in my manuscript unless I crop it manually.

```julia
function get_aspect(ax)
    if ax.aspect.val ∈ (DataAspect(), nothing)
        xmin, xmax = ax.xaxis.attributes[:limits].val
        ymin, ymax = ax.yaxis.attributes[:limits].val
        return (ymax - ymin) / (xmax - xmin)
    else
        return ax.aspect.val
    end
end

inch = 96
pt = 4 / 3
function fancy_savefig(fname, f, pass; fig_width=3.8inch, fig_height=3inch)
    if pass > 1
        aspect = get_aspect(f.current_axis.x)
        if aspect > desired_aspect
            colsize!(f.layout, 1, Aspect(1, 1 / aspect))
        else
            rowsize!(f.layout, 1, Aspect(1, aspect))
        end
    end
    resize_to_layout!(f)
    save(fname, f)
end

function mysave(fname, f)
   fancy_savefig(fname, f, 1)
   fancy_savefig(fname, f, 2)
end
```

The crazy thing is that it wasn't working by just going through the process once. After much trial and error, I found that it would only work properly if the figure was saved twice. Unfortunately, this 'solution' doesn't even always work. Sometimes the `ax.aspect` isn't even available to query. Other times, the `ax.xaxis.attributes[:limits]` returns the default `(0, 10)` even though I've already plotted something well outside those bounds.

**Comment by jkrumbiegel** [comment]
I think an easier solution for your kind of issue would be to transfer the current size of the axis scene to its width and height settings, thereby making its size available to the layout. We can't always have the layout react directly to the aspect because the aspect in the resulting layout might again lead to a different size, etc., that could lead to cyclical behavior. So yes, it's a bit annoying.

```julia
f, ax, sc = scatter(randn(50) .* 3, randn(50), axis = (; aspect = DataAspect()), figure = (; backgroundcolor = :tomato))
```

without the following code that would lead to

<img width="596" height="447" alt="Image" src="https://github.com/user-attachments/assets/b425b1f0-fc18-4e74-9b10-8e4328e24886" />

```julia
Makie.update_state_before_display!(f) # if you haven't displayed, yet, to update axis limits etc
axwidth, axheight = widths(ax.scene.viewport[])
ax.width = axwidth
ax.height = axheight
resize_to_layout!(f)
f
```

<img width="597" height="268" alt="Image" src="https://github.com/user-attachments/assets/84a2fe19-bb7e-4f2e-8ccc-8b25c5ba7c11" />

Basically, I would only use the `Aspect` route in more complex figures or when some interactive resizing should work.

**Comment by avonmoll** [comment]
That is a good solution! Thanks for pointing that out. Maybe it should be included in the aspect tutorial or somewhere else in the documentation.
